Robust Race Condition Checking in the TestHarness

Hysrok opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->
Created a new class, and additional methods and member variables to be able to check for race conditions in a robust way, and report on where race conditions were encountered.  Closes #13186

Commit:
Somewhat Stable Updates to the Robust Race Checker

Commit:
Somewhat Stable Updates to the Robust Race Checker

Commit:
Functional robust race checker added.  Tells where race conditions exist when run. This commit lacks adding the python tests and creating diagnostic solutions.

Commit:
Functional robust race checker added.  Tells where race conditions exist when run. This commit lacks adding the python tests and creating diagnostic solutions.

Commit:
fixup! Somewhat Stable Updates to the Robust Race Checker

Commit:
fixup! Somewhat Stable Updates to the Robust Race Checker

Commit:
Finished creating the robust race checker, ith it's tests. Put race checker into it's own class which the TestHarness class calls.

Commit:
Finished creating the robust race checker, ith it's tests. Put race checker into it's own class which the TestHarness class calls.

Commit:
Added the proper issue number to the race_condition test.

Commit:
Added the proper issue number to the race_condition test.

Commit:
Various change removing unused files and and fixing some style issues

Commit:
Created the robust race conditions checker.

- Added files to create and test the race condition checker.
- Edited current files to make use of the race condition checker.

closes #13186

Commit:
Merge branch 'feature_13186' of github.com:Hysrok/moose into feature_13186

permcody commented:
Hmm - Your git history is a little bit screwy here. I see a few unmerged "fixup" commits and a merge too?

Not the end of the world but you should be using "git rebase" to keep everything clean as you incorporate changes from the main repo (usually "upstream").

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -562,6 +564,102 @@ def cleanup(self):
             # Perform any write-to-disc operations
             self.writeResults()
 
+    def findRacePartners(self):
```
Shouldn't all three of these methods be offloaded to your RaceConditionChecker class too? The TestHarness should ideally just be querying the Scheduler, passing everything to the RaceConditionChecker and getting back results.

Hysrok replied:
Yeah, there should already be exact copies in the RaceConditionChecker class.  I thought I had removed this code from here, but I think I accidentally reverted the commit or undid it before committing or something like that.  I'll remove it remove this code and make sure it still works.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/343456/) on a6970f4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13474/site)

This comment will be updated on new commits.

permcody reviewed (commented):
Looks really good - Nice work!

permcody commented:
FYI - @milljm. This looks pretty good - you will need to work with Sam to get your PR in with his.

milljm commented:
@permcody I _thought_ my PR #13424 would conflict with this one, but it does not.

I merged his branch with mine, no issues. Python unittests also pass...

milljm commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -82,6 +95,42 @@ def __init__(self, tester, job_dag, options):
+        t_list = []
+        for temp_job in somejobs.all_downstreams(a_job):
+            t_list.append(temp_job.getTestName())
+        return t_list
```
I am little confused as to the purpose of DAG node management within the Job class. Is there a way we can move DAG operations into the JobDAG instead? Basically, I'm asking you if it is possible to instead create wrapper methods into the JobDAG object instead of doing the actual JobDAG work in the Job object. Something like:

JobDAG.py:
```python
def getUpstream(self, job):
    """ Method to return a list of all the jobs that need to be run before this job """
    t_list = []
        for temp_job in self.__job_dag.predecessors(job):
            t_list.append(temp_job.getTestName())
        return t_list

def getDownstreams(self, job):
    """ Method to return a list of all the jobs that need to be run after this job """
    t_list = []
    for temp_job in self.__job_dag.all_downstreams(job):
        t_list.append(temp_job.getTestName())
    return t_list
```

Job.py then looks like:
```python
def getUpstream(self):
    return self.__job_dag.getUpstream(self)

def getDownstreams(self):
    return self.__job_dag.getUpstream(self)
```

milljm replied:
Also... Cody and I have repeatedly not liked each other's naming convention for upstream/dependents/predecessors, etc. We really should take a vote on what to call this feature. getUpstream(C) does what exact?

A <-- B <-- C (as the DAG constructs it based on prereqs)

returns A and B?

But when we write the DAG on a white board like this:

A --> B --> C (a runs, then b runs, then c...)

what then should 'getUpstream' return? -this is where the confusion lies. To re-iterate: You are not doing anything incorrect. I would just like to call the vote on what to call these two procedures.

Hysrok replied:
Ok, I think anything referring to getting information about the JobDAG should be in the JobDAG, so these functions would be better in the JobDAG class with wrappers in the Job Class.  I was thinking though that each job itself has a unique set of predecessors, descendants, and race partners , so I was storing that unique set or list in the Job Class.  Is that alright?  In addition, when I tried to store the predecessor and descendant variable information in a DAG in the JobDAG, I found that they were getting deleted by the JobDAG's self-cleaning.

milljm replied:
I think storing this information in the Job class is a good idea. But performing the work... yeah that belongs elsewhere (in the JobDAG class).

Yeah, I got bitten by that as well one day. Is why I created getOriginalDAG in (in the dag object). While I peruse over that method, it is hacky. It is very difficult to understand when a DAG is finished, with no more incoming changes. In fact, my PR #13424 addresses this by removing as many places where it gets re-set as possible.

milljm commented on python/contrib/dag/__init__.py:
```diff
@@ -26,13 +27,65 @@ def __init__(self):
+        for fullyQualifiedFile in newTimes:
+            if not fullyQualifiedFile in originalTimes:
+                changed.append(fullyQualifiedFile)
+        return changed
```
I believe much of the above belongs in the JobDAG object. This code looks similar to the _doRaceConditions that exist the JobDAG object.

Unless you plan on manipulating the actual DAG object, it should probably exist elsewhere.

Hysrok replied:
I agree.  I wasn't sure where to put it, so I was working on it here when I was trying to approach the problem in a different way.  I think the JobDAG would be a great place for it, I'll move it right away.

Hysrok replied:
I actually decided to put it into its own file since it seems to be different enough from all the other methods in the JobDAG.  I made a new class called FileChecker which now has these methods.  Is that alright?

milljm replied:
Lets see it 👍

milljm commented on python/contrib/dag/__init__.py:
```diff
@@ -26,13 +27,65 @@ def __init__(self):
+                v = queue.pop()
+                self.add_edge(v, u)
+                queue.append(v)
+        return graph
```
👍

milljm commented on python/TestHarness/RaceChecker.py:
```diff
@@ -0,0 +1,112 @@
+                print("The following don't share any prereqs")
+            for job in colissions[prereq]:
+                print(job)
+        return
```
Not knowing when/how this Class gets called/used, we have a method you can make use of for printing information. See the following on how you can do that:
https://github.com/idaholab/moose/blob/a8876f99b2ea9983747eb05fde91873e42546524/python/TestHarness/JobDAG.py#L84

It will also help to use the above when RaceChecker is run in parallel; Any print operations will collide with other print operations. This is of course if your RaceChecker can run in parallel. I have not studied your PR close enough to see when it is called.

Hysrok replied:
The Class is called once everything is over, however, the printRaceConditionsByPrereq was something I was creating to better organize information that I thought we might want. I never ended up using it, and it doesn't seem to really add information that is different enough or more useful that what the printUniqueRacerSets does, so I'm removing that. Since the information isn't complete until everything has finished, none of this is called until the end when everything stops running in parallel. So I think it's fine?

milljm replied:
Regardless of when, or how, what ever you add to `job.setOutput()`, will be displayed back to the user, in a nice, formatted way. If you find you are spending any amount of time on "making the stdout look nice", then you should probably be using our print utility.

Hysrok replied:
Alright, I've been working on this, but I haven't made any progress yet.  Each time I think I understand it, I run some tests and nothing prints out.  I'll keep working on it.

milljm replied:
Most likely due to the fact RaceChecker is happening within a thread/child.
Try instead, to set some output using

```python
 job.setOutput("some output")
```

and see what happens.

Hysrok replied:
Yeah, I've actually been putting

    job.setOutput("something")

in a lot of places in the code.  I've got some in the TestHarness class, the JobDAG, and in the Job class (where it's self.setOutput) as well as in my RaceChecker class.  I've tried to use the printOutput function in the TestHarness as well as just using the getOutput function from the Job class, but it doesn't seem to be doing anything.

milljm replied:
One would set job.setOutput() to avoid having to use print statements. The TestHarness calls getOutput at the appropriate time, serially, to print results to the screen. If you are setting output, but not seeing that output, it's possible another class is overwriting that output. As I recall there is some sort of safety setting that prevents this... 

indeed, I see some interesting logic in Job.py: setOutput()

Hysrok replied:
So, I can see that setOutput is setting it, because I can call print(getOutput) and it prints it.  I think what's happening is that my code gets called after the TestHarness has already printed everything.  But my code needs to happen after everything, because that's when we know for sure which files have been created/changed.  So how should I address this?  Should I find how TestHarness prints everything and make it happen later?  Or should I just leave print statements?  Or is there another option?

Hysrok commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -171,6 +220,11 @@ def run(self):
         """
+        if self.options.testharness_diagnostics:
+            # Give some time so that file modified time has a distinct new value.
+            time.sleep(1)
```
Ok, I'm not sure exactly how to get this done otherwise.  The modified time goes in increments of 1 second.  If one job completes in under a second, then when it updates its modified time, it writes the same time that it picked it up at, so it doesn't record it as a file that has been written or edited.  I found code that was supposed to give me more precision on the modified time, but it didn't change anything.  Seems like the system only records up to the second for this Mac.  Suggestions?

milljm replied:
Do we know that these files _are_ created before we actually run (seriously, I am asking, because I don't know)? If we do, we might instead record the clock at which time the tester launched, using the-already-included clock() object (self.__start_time), and do a delta comparison with self.__end_time maybe?

Also, if we use those clock objects, you will save yourself a false-positive due to the possibility we exit this run method (due to self.__tester.shouldExecute() being False).

I am just brain-storming here...

Hysrok replied:
As far as I understand, some may be written before a test is run, while others won't exist until after the test runs.  How precise is the clock()?  Is it based on system time?  I guess I could look that up.

milljm replied:
For the files not created, and that we don't know about... yeah thats a toughy. But if we wanted to _assume_ anything created is going to be at clock start time... that might work. The internal clock is going to be precise enough. Try it out:

```python
from timeit import default_timer as clock
start_time = clock()
stop_time = clock()
start_time
stop_time
```
start_time and stop_time will be slightly different.

Hysrok replied:
So I'm trying to do this, but I can't quite get my head around logic that'll make it function.  Right now I'm using:
```suggestion
os.path.getmtime(fullyQualifiedFile)
```
to get the time it was modified.  I don't want to record a change in time if it wasn't modified, because the only way that I can tell if a file was changed is to compare the last-modified times.  Is there another way to flag if something was modified? The only way I had found was to to compare modified times, but if there is another way, I probably wouldn't even need to record times.

Hysrok replied:
Didn't mean to make that into a suggestion...  just meant to insert the code:
`os.path.getmtime(fullyQualifiedFile)`

Hysrok replied:
So while trying to fix this, I discovered a major bug.  Right now, if run on multiple cores, it is possible for for false positives to appear.  Looks like another job could potentially start it checking times before the previous job has finished changing the times.  So I was wondering is I should add a lock somewhere like in the RunParallel.py file:
`        if self.options.testharness_diagnostics:
            with self.locker:
                job.run()

        else:
            job.run()
`

Also, in order to avoid using sleep, I currently have:
`        if self.options.testharness_diagnostics:
            # Give some time so that file modified time has a distinct new value.
            before_check_time = clock()
            # Before the job does anything, get the times files below it were last modified
            ...
            after_check_time = clock()
            while(after_check_time - before_check_time < 1):
                after_check_time = clock()`

Commit:
Applied changes suggested from comments made in the pull request, as well as adding proper error codes for race conditions.
- Removed reduntant RaceChecker code from the TestHarness
- Moved some DAG management from Job to the JobDAG
- Created FileChecker class and moved applicable code from the DAG class 
to FileChecker
- Discovered major bug with jobs racing to update modified times

milljm commented on python/TestHarness/StatusSystem.py:
```diff
@@ -62,7 +63,7 @@ class StatusSystem(object):
     diff = status(status='DIFF', color='YELLOW', code=0x81)
     deleted = status(status='DELETED', color='RED', code=0x83)
-    error  = status(status='ERROR', color='RED', code=0x80)
+    error  = status(status='ERROR', color='RED', code=0x84)
```
0x80 was by design and not a typo. If we are trying to be more specialized, I have no problems with it. I am not using 'code' during logical status comparisons so I suppose this is alright.

I do notice you are attempting to add a new status. But I do not see a new status tuple?

Hysrok replied:
Ok, I was just making it reflect the part above where 'fail' has the 0x80 code and error has the code 0x84.  Figured two statuses using the same code would make using the code less useful for identifying the root cause or something like that.  If not, I would propose changing the above statement to read that the 'error' exit code is 0x80.

As for the status I was adding, I added a description and the exit code that I am using just as an informational reference.  As my code stands, it wouldn't be using the tuple, so I wasn't sure if I should add one that never got used.

milljm commented on python/TestHarness/FileChecker.py:
```diff
@@ -0,0 +1,48 @@
+import os
```
Include the copyright header

Hysrok replied:
Oops, I'll go ahead and add that now.

milljm commented on python/TestHarness/RaceChecker.py:
```diff
@@ -0,0 +1,112 @@
+                            job_a.racePartners.add(job_b)
+                            job_b.racePartners.add(job_a)
+                            raceConditionsExist = True
+        return raceConditionsExist
```
Okay, so I am stuck at this method. There are better ways of going about the nested for loops (use sets instead to perform differences/intersections), but the more important issue is, are we needing job_$.getDownstreamNodes()?

If we do, then I am unclear as to why. From what I understand about this PR, we have made every job have an edge pointing at each other. So theoretically, there should always be downstreams.

However, what is unexpected about it, is that you are using the DAG which gets deconstructed as it passes through the scheduler. So there are never any downstream jobs by the time we get this far. I think We need to use the dag.getOriginalDAG's DAG to perform this operation correctly.

While I have been toying with this method using the original dag, I am still unclear as to when we want to actually detect a race condition.

Some other optimizations this method needs, is whether or not to perform additional work if certain conditions are met (effectively does the same thing)
```python
def findRacePartners(self):
    for job_a, job_b in itertools.combinations(job_group, 2):
        # Job was skipped/never ran
        if job_a.isSkip() or job_b.isSkip():
            continue
        # No matching files
        elif not set(job_a.modifiedFiles).intersection(set(job_b.modifiedFiles)):
            continue
        # job_a and job_b have similar downstream jobs
        elif not set(job_a.getDownstreamNodes()).intersection(set(job_b.getDownstreamNodes())):
            job_a.racePartners.add(job_b)
            job_b.racePartners.add(job_a)
            return True
```
Even though I have optimized this method, I don't think it is correct. But I wanted to give you an idea how it _could_ be optimized.

milljm replied:
As we discussed, I thought we were operating on the DAG when doing job_a.getDownstreamNodes(). My bad. Rather than edit my comment, I'm replying so as to preserve some history. Some of it is useful :)

permcody commented:
Closing in favor of #13535.
